src/gcp_bigquery.rs

```diff
@@ -0,0 +1,449 @@
+}
+
+#[derive(Debug)]
+pub enum GcpClientError {
+    MissingParameters,
+    Other(String),
+    BigQueryError(BQError),
+}
+
+///
+/// Factory function to initialize the GCP bigquery client
+/// Also deletes / creates tables as necessary (controlled via setting)
+///
+/// # Arguments
+///
+/// * `indexer_gcp_bigquery_config` - gcp config block
+/// * `event_mappings` - list of event mapping blocks (all events to parse)
+pub async fn init_gcp_bigquery_db(
+    indexer_gcp_bigquery_config: &IndexerGcpBigQueryConfig,
+    event_mappings: &[IndexerContractMapping],
+) -> Result<GcpBigQueryClient, GcpClientError> {
+    let gcp_bigquery = GcpBigQueryClient::new(&indexer_gcp_bigquery_config, &event_mappings)
+        .await
+        .unwrap();
+
+    gcp_bigquery.delete_tables().await;
+    gcp_bigquery.create_tables().await;
+
+    Ok(gcp_bigquery)
+}
+
+///
+/// This function merges the common columns specified in COMMON_COLUMNS with the
+/// columns specified in the reth-indexer-config file for the particular event type to parse
+/// The types are ordered to allow for consistent column order within the GCP tables
+/// This is done for each table specified in the indexer
+///
+/// # Arguments
+///
+/// * `indexer_event_mappings` - list of all event mappings
+pub fn load_table_configs(
+    indexer_event_mappings: &[IndexerContractMapping],
+) -> HashMap<String, IndexMap<String, String>> {
+    //  Config map for all tables
+    //  Ordered by common columns, then by configured mapped
+    let mut all_tables: HashMap<String, IndexMap<String, String>> = HashMap::new();
+
+    for mapping in indexer_event_mappings {
+        for abi_item in mapping.decode_abi_items.iter() {
+            let table_name = abi_item.name.to_lowercase();
+            let column_type_map: IndexMap<String, String> = abi_item
+                .inputs
+                .iter()
+                .map(|input| {
+                    (
+                        input.name.clone(),
+                        solidity_type_to_db_type(input.type_.clone().as_str()).to_string(),
+                    )
+                })
+                .collect();
+
+            let common_column_type_map: IndexMap<String, String> = COMMON_COLUMNS
+                .into_iter()
+                .map(|it| (it.0.to_string(), it.1.to_string()))
+                .collect();
+
+            let merged_column_types: IndexMap<String, String> = common_column_type_map
+                .into_iter()
+                .chain(column_type_map)
+                .collect();
+
+            all_tables.insert(table_name, merged_column_types);
+        }
+    }
+
+    all_tables
+}
+
+///
+/// Maps solidity types (in indexer config) to bigquery column types
+/// The configuration / reth uses solidity types, we map these to types that we'll use in GCP
+/// bigquery
+///
+/// # Arguments
+///
+/// * `abi_type` - the ABI type, specified as a string
+fn solidity_type_to_db_type(abi_type: &str) -> &str {
```

Comment:
can these not now be shared between postgres stuff in a helper file? save repeated code?

Reply:
missed this comment?

Reply:
the name of the function is the same between the implementations, but the returned type string is different.  the postgres returns CHAR(42), TEXT, or NUMERIC (the actual sql types used in the table definition), whereas the gcp one returns (string or int), which are just saved internally in a type representation so that it can get converted to TableFieldSchema::integer and TableFieldSchema::string in the table creation statement.   if we were to merge together, the shared function would have to be implementation-aware?   i felt like it was ok to keep it simple and isolated to each respective module, even if the name of the function is present in both.